contracts/CommunityToken.sol

```diff
@@ -0,0 +1,148 @@
+    address communityTokenRewardFactoryAddress;
+  }
+
+  struct CommunityTokenContainer {
+    CommunityTokenInfo info;
+    mapping(uint16 => RewardPeriodParams) rewardPeriodParams;    // period
+  }
+
+  struct RewardPeriodParams {
+    uint256 maxTotalTokenPool;
+    uint256 maxRewardPerUser;
+    uint256 availableBalance;
+    uint256 totalTokenPool;
+
+    mapping(address => bool) isClaimed;
+  }
+
+  CommunityTokenContainer communityTokenContainer;
+
+  event CommunityRewardSettingsUpdated();
+
+  constructor(address tokenAddress, uint256 maxRewardPerPeriod, uint256 maxRewardPerUser, address communityTokenRewardFactoryAddress) {
+    // if address = 0?
+    communityTokenContainer.info.tokenAddress = tokenAddress;
+    communityTokenContainer.info.name = IERC20Metadata(communityTokenContainer.info.tokenAddress).name();
+    communityTokenContainer.info.symbol = IERC20Metadata(communityTokenContainer.info.tokenAddress).symbol();
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerUser;
+    communityTokenContainer.info.createTime = CommonLib.getTimestamp();
+    communityTokenContainer.info.communityTokenRewardFactoryAddress = communityTokenRewardFactoryAddress;
+  }
+
+  // This is to support Native meta transactions
+  // never use msg.sender directly, use _msgSender() instead
+  function _msgSender()
+      internal
+      override
+      view
+      returns (address sender)
+  {
+    return NativeMetaTransaction._msgSender();
+  }
+
+  function updateCommunityRewardSettings(uint256 maxRewardPerPeriod, uint256 maxRewardPerUser) external override { // check role
+    // dispatcherCheck(userAddress); 
+    
+    communityTokenContainer.info.maxRewardPerPeriod = maxRewardPerPeriod;
+    communityTokenContainer.info.maxRewardPerUser = maxRewardPerUser;
+    emit CommunityRewardSettingsUpdated();
+  }
+
+  function getAvailableRewardsBalance() public view override returns(uint256) {   // public?
+    uint256 balance = IERC20(communityTokenContainer.info.tokenAddress).balanceOf(address(this));
+    if (balance <= communityTokenContainer.info.reservedTokens) {
+      return 0;
+    }
+
+    return balance - communityTokenContainer.info.reservedTokens;
+  }
+
+  // get pool
+  function getTotalPeriodReward(uint16 period) public view returns(uint256 maxTotalTokenPool, uint256 maxTokensPerUser, uint256 balance) {
+    RewardPeriodParams storage rewardPeriodParams = communityTokenContainer.rewardPeriodParams[period];
+    return (rewardPeriodParams.maxTotalTokenPool, rewardPeriodParams.maxRewardPerUser, rewardPeriodParams.availableBalance);
+  }
+
+  // set pool
+  function setReadyToClaimPeriodRewards(RewardLib.PeriodRewardShares memory periodRewardShares, uint16 period) external override {
+    require(_msgSender() == communityTokenContainer.info.communityTokenRewardFactoryAddress, "only_community_token_reward_factory_contract_can_call_this_action");
```

Comment:
Verify that this action has not been executed already for this period. Don't allow to run it multiple times